### PR TITLE
Feat/ed25519 trial

### PR DESCRIPTION
**Pull request by raj-shubham**
Currently, this request is for a non-verifiable Shamir share.

### CODE REVIEW

File: ed25519/shamir.go [hunk 1]
```diff
@@ -1 +1,183 @@
...
+// polynomial with constant term equal to the given secret.
+//
+// Panics: This function will panic if k is greater than len(coeffs).
+func setRandomCoeffs(coeffs []Scalar, secret Scalar, k int) {
+	coeffs = coeffs[:k]
+	coeffs[0] = secret
+
+	// NOTE: If k > len(coeffs), then this will panic when i > len(coeffs).
+	for i := 1; i < k; i++ {
+		coeffs[i] = RandomScalar() // to be written
```
Comment (ross-pure) [inline ed25519/shamir.go:133]: Outdated comment.

File: ed25519/shamir.go [hunk 2]
```diff
@@ -1 +1,183 @@
...
+	}
+}
+
+// Evaluates the polynomial defined by the given coefficients at the point x
+// and stores the result in y. Modifies y, but leaves x and coeffs unchanged.
+// Normalizes y, so this this is not neccesary to do manually after calling
+// this function.
```
Comment (ross-pure) [inline ed25519/shamir.go:140]: Normalisation is only relevent for secp256k1 types due to their internal representation.
```suggestion
// Evaluates the polynomial defined by the given coefficients at the point x
// and stores the result in y. Modifies y, but leaves x and coeffs unchanged.
```

File: ed25519/shamirutil/misc.go [hunk 3]
```diff
@@ -0,0 +1,26 @@
+package shamirutil
```
Comment (ross-pure) [inline ed25519/shamirutil/misc.go:1]: This file (misc.go) probably doesn't need to be copied over. All of the functions in this file are exactly the same.

File: ed25519/shamir_test.go [hunk 4]
```diff
@@ -0,0 +1,385 @@
+package ed25519_test
+
+import (
+	"math/rand"
+	"testing"
+	"time"
+
+
```
Comment (ross-pure) [inline ed25519/shamir_test.go:8]: go format would make this just one space, not two.

**Review by ross-pure — COMMENTED** [review]
Looking good so far, but left some comments for changes to be made.

File: ed25519/shamir_test.go (hunk 4 above)
Comment (raj-shubham) [reply ed25519/shamir_test.go]: Executed go fmt.

File: ed25519/shamirutil/misc.go (hunk 3 above)
Comment (raj-shubham) [reply ed25519/shamirutil/misc.go]: Removed the misc.go file. Using the misc code from secp256k1.

File: ed25519/scalar.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	for _, v := range b {
+		if v != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// sets a uint16 value to a scalar
+func (s *Scalar) SetU16(i uint16) {
```
[the remark this answers is not in the document]
Comment (raj-shubham) [reply ed25519/scalar.go]: Created a SetU32 function. Do we need SetU64?

File: ed25519/scalar.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
+package ed25519
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"filippo.io/edwards25519"
+	"fmt"
+	"github.com/renproject/surge"
+	"unsafe"
+)
```
[the remark this answers is not in the document]
Comment (ross-pure) [reply ed25519/scalar.go:10]: Ah, turns out the command that formats the imports is `goimports`. Running that changes the order/spacing.

**Comment by ross-pure** [comment]
LGTM after recent changes

**Pull request merged**